### PR TITLE
Relax zstd-sys Version Pin

### PR DESCRIPTION
**Pull request by waynexia**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

I hit this while trying to upgrade arrow for datafusion.

Parquet depends on `zstd-sys` and datafusion depends on `async-compression` which then depends on `zstd-sys` via `zstd-safe`.

Parquet cannot use `zstd-sys` beyond `2.0.9` due to https://github.com/gyscos/zstd-rs/issues/269. And `zstd-safe` has two available versions `7.0.0` which depends on `zstd-sys` `2.0.7` and `7.1.0` on `zstd-sys` `2.10.0`. 

So to make parquet be compatible with projects use `async-compression`, the only available version to `zstd-sys` is `2.0.7`

# What changes are included in this PR?

Downgrade and pin `zstd-sys` to `2.0.7`

# Are there any user-facing changes?

Yes, the dependences

This patch needs to be included in the next release https://github.com/apache/arrow-rs/issues/5688 if possible

### CODE REVIEW

**Comment by kylebarron** [comment]
Maybe <= 2.0.9?

**Comment by waynexia** [comment]
> Maybe <= 2.0.9?

That looks better 👍 Changed in [fc4c06a](https://github.com/apache/arrow-rs/pull/5829/commits/fc4c06a96c1e6e8d888ca379ea0283f32f3f336b)

File: parquet/Cargo.toml [hunk 1]
```diff
@@ -55,7 +55,7 @@ lz4_flex = { version = "0.11", default-features = false, features = ["std", "fra
 zstd = { version = "0.13.0", optional = true, default-features = false }
 # TODO: temporary to fix parquet wasm build
 # upstream issue: https://github.com/gyscos/zstd-rs/issues/269
-zstd-sys = { version = "=2.0.7", optional = true, default-features = false }
+zstd-sys = { version = "<=2.0.9", optional = true, default-features = false }
```
Comment (Xuanwo) [inline parquet/Cargo.toml]: This could pick up `zstd-sys 1.0`, please make sure we also have `>= 2.0.0`.

**Comment by waynexia** [comment]
Well, it turns out that cargo does not support "less than" comparison on patch position. The patch version is ignored in `<2.0.10` and leads to a selection on `1.x`.

>package `zstd-sys v1.4.3+zstd.1.3.4`
    ... which satisfies dependency `zstd-sys = "<2.0.10"` (locked to 1.4.3+zstd.1.3.4) of package `parquet v51.0.0 (/__w/arrow-rs/arrow-rs/parquet)` links to native library `zstd`

I'm afraid we have to use `=2.0.7` instead...

**Comment by tustvold** [comment]
I'm somewhat inclined to think we should just remove the version pin, and document that people using wasm32 will need to manually pin zstd-sys themselves. We're creating pain for all of our users by pinning this crate, to save some pain for a very small subset of our community...

File: parquet/Cargo.toml (hunk 1 above)
Comment (waynexia) [reply parquet/Cargo.toml]: Ahh, that works 💯

**Comment by Xuanwo** [comment]
> I'm somewhat inclined to think we should just remove the version pin, and document that people using wasm32 will need to manually pin zstd-sys themselves.

I agree that libs like parquet should not pin dependencies. This forces users to adopt the same pinned version, increasing the likelihood of conflicts. Instead of pinning versions downstream, it's better for us to collaborate upstream to address the issue.

**Comment by waynexia** [comment]
> I'm somewhat inclined to think we should just remove the version pin, and document that people using wasm32 will need to manually pin zstd-sys themselves. We're creating pain for all of our users by pinning this crate, to save some pain for a very small subset of our community... Ultimately if zstd-sys broke support for wasm32 it is on them to fix it

That's also an option. Cargo provides [patch mechanism](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#overriding-dependencies) to override deps

**Comment by tustvold** [comment]
> Cargo provides [patch mechanism](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#overriding-dependencies) to override deps

They shouldn't need to go that far, if they have a lockfile they can just use `cargo update --precise`, otherwise they can just declare an explicit dependency on zstd-sys and pin it there.

File: parquet/Cargo.toml [hunk 2]
```diff
@@ -52,10 +52,8 @@ snap = { version = "1.0", default-features = false, optional = true }
 brotli = { version = "6.0", default-features = false, features = ["std"], optional = true }
 flate2 = { version = "1.0", default-features = false, features = ["rust_backend"], optional = true }
 lz4_flex = { version = "0.11", default-features = false, features = ["std", "frame"], optional = true }
-zstd = { version = "0.13.0", optional = true, default-features = false }
-# TODO: temporary to fix parquet wasm build
-# upstream issue: https://github.com/gyscos/zstd-rs/issues/269
-zstd-sys = { version = "=2.0.9", optional = true, default-features = false }
+zstd = { version = "0.13", optional = true, default-features = false }
+zstd-sys = { version = "2.0", optional = true, default-features = false }
```
Comment (tustvold) [inline parquet/Cargo.toml:56]: I suspect zstd-sys can be removed entirely

**Review by tustvold — COMMENTED** [review]
You will need to update the wasm CI to pin zstd-sys

**Comment by waynexia** [comment]
Removed them in [d7d5e9b](https://github.com/apache/arrow-rs/pull/5829/commits/d7d5e9b6c8f46980dca809ddba8468c5128dfe4c) based on above discussion

File: parquet/Cargo.toml (hunk 2 above)
Comment (waynexia) [reply parquet/Cargo.toml]: Moved to target-specified deps (for CI)

**Comment by waynexia** [comment]
Thank you all :heart:

**Pull request merged**